examples/ex71.cpp

```diff
@@ -0,0 +1,385 @@
+   ConstantCoefficient load(1.000000000);
+   ConstantCoefficient c_ee(0.000000001);
+
+   // 5. Define the finite element spaces for the solution
+   H1_FECollection fec(order, dim);
+   FiniteElementSpace fespace(mesh, &fec, 1, Ordering::byVDIM);
+   int glob_size = fespace.GetTrueVSize();
+
+   std::cout << "Number of finite element unknowns: " << glob_size << std::endl;
+
+   // 6. Define the essential boundary attributes
+   Array<int> ess_bdr(mesh->bdr_attributes.Max());
+   ess_bdr = 1;
+
+   // 7. Define the nonlinear form
+   NonlinearForm *nf = new NonlinearForm(&fespace);
+
+   // 8. Define the solution vector x
+   GridFunction x(&fespace);
+   x = 0.0;
+   Vector tv(fespace.GetTrueVSize());
+   Vector sv(fespace.GetTrueVSize());
+   tv = 0.0;
+   sv = 0.0;
+
+   // 9. Define ParaView DataCollection
+   ParaViewDataCollection *dacol = new ParaViewDataCollection("Example71", mesh);
+   dacol->SetLevelsOfDetail(order);
+   dacol->RegisterField("sol", &x);
+
+   // 11. Set domain integrators - start with linear diffusion
+   {
+      // The default power coefficient is 2.0
+      ConstantCoefficient lpp(2.0);
+      if (integrator == 0)
+      {
+         nf->AddDomainIntegrator(new pLaplace(lpp, c_ee, load));
+      }
+      else if (integrator == 1)
+      {
+         nf->AddDomainIntegrator(new pLaplaceAD(lpp, c_ee, load));
+      }
+      nf->SetEssentialBC(ess_bdr);
+
+      // Compute the energy
+      double energy = nf->GetEnergy(tv);
+      std::cout << "[2] The total energy of the system is E=" << energy
+                << std::endl;
+
+      // Time the assembly
+      timer->Clear();
+      timer->Start();
+      nf->GetGradient(sv);
+      timer->Stop();
+      std::cout << "[2] The assembly time is: " << timer->RealTime()
+                << std::endl;
+
+      Solver *prec;
+#ifdef MFEM_USE_SUITESPARSE
+      prec = new UMFPackSolver();
+#else
+      prec = new GSSmoother();
+#endif
+      CGSolver *j_pcg = new CGSolver();
+      j_pcg->SetRelTol(1e-7);
+      j_pcg->SetAbsTol(1e-15);
+      j_pcg->SetMaxIter(500);
+      j_pcg->SetPrintLevel(print_level);
+      j_pcg->SetPreconditioner(*prec);
+
+      NewtonSolver *ns;
+      ns = new NewtonSolver();
+      ns->iterative_mode = true;
+      ns->SetSolver(*j_pcg);
+      ns->SetOperator(*nf);
+      ns->SetPrintLevel(print_level);
+      ns->SetRelTol(1e-6);
+      ns->SetAbsTol(1e-12);
+      ns->SetMaxIter(10);
+
+      // Solve the problem
+      timer->Clear();
+      timer->Start();
+      ns->Mult(tv, sv);
+      timer->Stop();
+      std::cout << "Time for the NewtonSolver: " << timer->RealTime()
+                << std::endl;
+
+      energy = nf->GetEnergy(sv);
+      std::cout << "[pp=2] The total energy of the system is E=" << energy
+                << std::endl;
+
+      delete ns;
+      delete j_pcg;
+      delete prec;
+
+      x.SetFromTrueDofs(sv);
+      dacol->SetTime(2.0);
+      dacol->SetCycle(2);
+      dacol->Save();
+   }
```

Comment:
This region is executed multiple times in the same way. Please make this a function and call it from the loops or special cases.

Reply:
extracted and separated in a class